### ISSUE TITLE
[ALSA] Enumerate example panics in cpal::cpal_impl::Device::supported_formats

### ISSUE THREAD

**Issue by mateon1** [tomaka/cpal#206]
Trying to run the enumerate example causes a panic:

```gdb
mateon@hydra:/shared/dev/rust/cpal$ RUST_BACKTRACE=1 cargo run --example enumerate
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/examples/enumerate`
Default Input Device:
  Some("default")
Default Output Device:
  Some("default")
Devices: 
ALSA lib pcm_dmix.c:1052:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:1052:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
1. "iec958:CARD=PCH,DEV=0"
Error: DeviceNotAvailable
ALSA lib pcm_dmix.c:1052:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
2. "dmix:CARD=PCH,DEV=1"
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
thread 'main' panicked at 'device not available: "Invalid argument"', libcore/result.rs:945:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:380
   3: std::panicking::default_hook
             at libstd/panicking.rs:396
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:576
   5: std::panicking::begin_panic
             at libstd/panicking.rs:537
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:521
   7: rust_begin_unwind
             at libstd/panicking.rs:497
   8: core::panicking::panic_fmt
             at libcore/panicking.rs:71
   9: core::result::unwrap_failed
             at /checkout/src/libcore/macros.rs:23
  10: <core::result::Result<T, E>>::expect
             at /checkout/src/libcore/result.rs:809
  11: cpal::cpal_impl::Device::supported_formats
             at src/alsa/mod.rs:95
  12: cpal::cpal_impl::Device::default_format
             at src/alsa/mod.rs:270
  13: cpal::cpal_impl::Device::default_input_format
             at src/alsa/mod.rs:296
  14: cpal::Device::default_input_format
             at ./src/lib.rs:388
  15: enumerate::main
             at examples/enumerate.rs:15
  16: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
  17: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:479
  18: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102
  19: std::rt::lang_start_internal
             at libstd/panicking.rs:458
             at libstd/panic.rs:358
             at libstd/rt.rs:58
  20: std::rt::lang_start
             at /checkout/src/libstd/rt.rs:74
  21: main
  22: __libc_start_main
  23: _start
```

Both other examples fail as well, at `default_device.default_output_format().expect(...)`

**Comment by mateon1** [comment]
By modifying `src/alsa/mod.rs` at line 95, I found that the returned error code is `-22`.
Adding it to the `return Err` path lets the example finish with the following output

<details>

```rust
mateon@hydra:/shared/dev/rust/cpal$ RUST_BACKTRACE=1 cargo run --example enumerate
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/examples/enumerate`
Default Input Device:
  Some("default")
Default Output Device:
  Some("default")
Devices: 
ALSA lib pcm_dmix.c:1052:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:1052:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
1. "iec958:CARD=PCH,DEV=0"
Error: DeviceNotAvailable
ALSA lib pcm_dmix.c:1052:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
2. "dmix:CARD=PCH,DEV=1"
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Error: DeviceNotAvailable
ALSA lib pcm_dsnoop.c:556:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dsnoop.c:556:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
3. "hw:CARD=PCH,DEV=1"
Error: DeviceNotAvailable
4. "plughw:CARD=PCH,DEV=1"
Error: DeviceNotAvailable
5. "hdmi:CARD=HDMI,DEV=0"
Error: DeviceNotAvailable
6. "hdmi:CARD=HDMI,DEV=1"
Error: DeviceNotAvailable
7. "hdmi:CARD=HDMI,DEV=2"
Error: DeviceNotAvailable
8. "hdmi:CARD=HDMI,DEV=3"
Error: DeviceNotAvailable
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
9. "dmix:CARD=HDMI,DEV=3"
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Error: DeviceNotAvailable
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
10. "dmix:CARD=HDMI,DEV=7"
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Error: DeviceNotAvailable
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
11. "dmix:CARD=HDMI,DEV=8"
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Error: DeviceNotAvailable
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
12. "dmix:CARD=HDMI,DEV=9"
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
Error: DeviceNotAvailable
ALSA lib pcm_dsnoop.c:556:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dsnoop.c:556:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dsnoop.c:556:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dsnoop.c:556:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:618:(snd_pcm_dsnoop_open) unable to open slave
13. "hw:CARD=HDMI,DEV=3"
Error: DeviceNotAvailable
14. "hw:CARD=HDMI,DEV=7"
Error: DeviceNotAvailable
15. "hw:CARD=HDMI,DEV=8"
Error: DeviceNotAvailable
16. "hw:CARD=HDMI,DEV=9"
Error: DeviceNotAvailable
17. "plughw:CARD=HDMI,DEV=3"
Error: DeviceNotAvailable
18. "plughw:CARD=HDMI,DEV=7"
Error: DeviceNotAvailable
19. "plughw:CARD=HDMI,DEV=8"
Error: DeviceNotAvailable
20. "plughw:CARD=HDMI,DEV=9"
Error: DeviceNotAvailable
```
</details>

If I'm reading that correctly, it means no sound devices are available...

**Comment by mitchmindtree** [comment]
Thanks for the report!

> By modifying src/alsa/mod.rs at line 95, I found that the returned error code is -22.
Adding it to the return Err path lets the example finish

Is there any chance you could submit a PR for this?

> If I'm reading that correctly, it means no sound devices are available...

What devices do you expect  to be available on your system? Note that ALSA will not consider a device available if it is currently in use by another application (this is why most linux applications target pulseaudio or jack instead). If you close all other audio applications and then run `cargo run --example enumerate` do you get different results?

It looks like the `device not available: "Invalid argument"` error you were getting is from [this line](https://github.com/tomaka/cpal/blob/master/src/alsa/mod.rs#L95) - I wonder if by invalid argument it is referring to the stream non supporting `SND_PCM_NONBLOCK` or if some devices return this error when attempting open them as a capture device when they only support playback (or vice versa).

**Comment by azyobuzin** [comment]
It looks that this issue and #227 is the same problem.

>  I wonder if by invalid argument it is referring to the stream non supporting SND_PCM_NONBLOCK or if some devices return this error when attempting open them as a capture device when they only support playback (or vice versa).

In my environment, opening dmix device returns -22 if I specify `SND_PCM_STREAM_CAPTURE`.

```rust
extern crate alsa_sys as alsa;
use std::mem;

fn main() {
    const TARGET_DEVICE: &'static str = "dmix:CARD=PCH,DEV=0";
    let name_zeroed = std::ffi::CString::new(TARGET_DEVICE).unwrap();

    println!("Open as a playback stream");
    unsafe {
        let mut handle = mem::uninitialized();
        let result = alsa::snd_pcm_open(
            &mut handle,
            name_zeroed.as_ptr() as *const _,
            alsa::SND_PCM_STREAM_PLAYBACK,
            alsa::SND_PCM_NONBLOCK
        );
        println!("{}", result);
        if result == 0 { alsa::snd_pcm_close(handle); }
    }

    println!("Open as a capture stream");
    unsafe { 
        let mut handle = mem::uninitialized();
        let result = alsa::snd_pcm_open(
            &mut handle,
            name_zeroed.as_ptr() as *const _,
            alsa::SND_PCM_STREAM_CAPTURE,
            alsa::SND_PCM_NONBLOCK
        );
        println!("{}", result);
        if result == 0 { alsa::snd_pcm_close(handle); }
    }
}
```

stdout

```
Open as a playback stream
0
Open as a capture stream
-22
```

stderr

```
ALSA lib pcm_dmix.c:990:(snd_pcm_dmix_open) The dmix plugin supports only playback stream
```

```
$ uname -a
Linux azyobuzin-S300CA 4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

So, I think we can treat error code -22 as the device has no supported format in the specified data flow (playback or capture), like https://github.com/azyobuzin/cpal/commit/02e25e77aca235b0246b41ca837cb73dd6e795e7.

**Comment by mitchmindtree** [comment]
This should be closed via #248 and #286.
